pandas/io/sas/saslib.pyx

```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
 
     return np.asarray(outbuff).tostring()
 
-def process_byte_array_with_data(parser, int offset, int length, np.ndarray[uint8_t, ndim=2] byte_chunk,
-                                 np.ndarray[dtype=object, ndim=2] string_chunk):
+
+def _do_read(parser, int nrows):
+    cdef int i
+
+    for i in range(nrows):
+        done = _readline(parser)
+        if done:
+            break
+
+
+def _readline(parser):
+
+    cdef int offset
+    cdef int bit_offset
+    cdef int align_correction
+    cdef int subheader_pointer_length
+
+    bit_offset = parser._page_bit_offset
+    subheader_pointer_length = parser._subheader_pointer_length
+
+    # If there is no page, go to the end of the header and read a page.
+    if parser._cached_page is None:
+        parser._path_or_buf.seek(parser.header_length)
+        done = parser._read_next_page()
+        if done:
+            return True
+
+    # Loop until a data row is read
+    while True:
+        if parser._current_page_type == const.page_meta_type:
+            flag = (parser._current_row_on_page_index >=
```

Comment:
in order to get some real perf increases. you are going to actually have to keep these variables as typed in cython.